CONTRIBUTING.md

```diff
@@ -98,8 +98,82 @@ object provides an interface to a running Prometheus instance. This
 interface is limited to only those aspects of Prometheus required by
 this charm.
 
+### Library Details
+
+The Prometheus charm library exposes a provider and consumer object -
+`PrometheusProvider` and `PrometheusConsumer` along with the custom
+charm event `TargetsChanged` within the `MonitoringEvents` event
+descriptor. `PrometheusProvider` emits the `TargetsChagned` event in
+response to relation changed and departed events. It is expected that
+the provider charm would respond to this event and regenerate the
+Prometheus configuration using information provided by the `jobs()`
+and `alerts()` methods of the `PrometheusProvider`.
+
+The `jobs()` method gathers a list of scrape jobs from all related
+consumer charms. In doing so it invokes the `_static_scrape_config()`
+method for each such relation. `_static_scrape_config()` generates a
+unique job name for each job, sanitizes the job and associates Juju
+topology labels with it. Labeling of Juju topology is done by
+`_labeled_static_job_config()`. However the labeling differs for
+scrape targets whose host address was automatically gathered and those
+were addresses were explicitly specified. Hence three separate
+functions are required for labeling `_set_juju_labels()` for labels
+common to both but `_labeled_unitless_config()` and
+`_labeled_unit_config()` for the differing labels. The list of
+automatically gathered host address is provided by
+`_relation_hosts()`.
+
+The `PrometheusConsumer` is responsible for forwarding scrape
+configuration, scrape target addresses, scrape metadata and alert
+rules to the Prometheus provider. In doing so it also ensures that the
+alert rules have Juju topology labels and filters injected into
+them. The `_set_unit_ip()` methods forwards scrape target host
+addresses using unit relation data. The `_set_scrape_metadata()`
+method forwards all the other information using application relation
```

Comment:
> all the other information

consider clarifying

Reply:
The first sentence in the paragraph lists all the information forwarded. Let me know if this does not make it clear.